Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/MidReader.cs

```diff
@@ -416,142 +438,61 @@ private static void ReadNotes(IList<MidiEvent> track, Song song, Song.Instrument
+                    if (psEvent.Type != MidIOHelper.SYSEX_TYPE_PHRASE)
                     {
-                        uint tick = (uint)se1.AbsoluteTime;
-                        Song.Difficulty difficulty;
-                        switch (bytes[5])
-                        {
-                            case 0: difficulty = Song.Difficulty.Easy; break;
-                            case 1: difficulty = Song.Difficulty.Medium; break;
-                            case 2: difficulty = Song.Difficulty.Hard; break;
-                            case 3: difficulty = Song.Difficulty.Expert; break;
-                            default: continue;
-                        }
+                        Debug.LogWarning($"Encountered unknown Phase Shift SysEx event type {psEvent.Type}");
+                        continue;
+                    }
 
-                        uint endPos = 0;
-                        for (int j = i; j < tapAndOpenEvents.Count; ++j)
+                    if (psEvent.Value == MidIOHelper.SYSEX_VALUE_PHRASE_START)
+                    {
+                        sysexEventQueue.Add(new PhaseShiftSysExStart(psEvent));
+                    }
+                    else if (psEvent.Value == MidIOHelper.SYSEX_VALUE_PHRASE_END)
+                    {
+                        // Iterate through event queue in reverse to find the corresponding start event
+                        bool foundMatch = false;
+                        for (int startIndex = sysexEventQueue.Count - 1; startIndex >= 0; startIndex--)
                         {
-                            var se2 = tapAndOpenEvents[j] as SysexEvent;
-                            if (se2 != null)
+                            var startEvent = sysexEventQueue[startIndex];
+                            if (startEvent.AbsoluteTime <= psEvent.AbsoluteTime && startEvent.MatchesWith(psEvent))
                             {
-                                var b2 = se2.GetData();
-                                if (b2.Length == 8 && b2[5] == bytes[5] && b2[7] == 0)
-                                {
-                                    endPos = (uint)(se2.AbsoluteTime - tick);
+                                sysexEventQueue.RemoveAt(startIndex);
+                                startEvent.EndEvent = psEvent;
+                                foundMatch = true;
 
-                                    if (endPos > 0)
-                                        --endPos;
-
-                                    break;
+                                processParams.midiEvent = startEvent;
+                                EventProcessFn processFn;
+                                if (processParams.sysexProcessMap.TryGetValue(psEvent.Code, out processFn))
+                                {
+                                    processFn(processParams);
                                 }
+                                break;
                             }
                         }
```

Comment:
Neater to move this search into it's own function
`static bool TryLocateStartEvent(IList<PhaseShiftSysExStart> sysexEventQueue, PhaseShiftSysEx psEndEvent, out int startEventIndex)`
And then remove from the list perform the process function if that returns true, which should be overall more readable and avoids the RemoveAt in the middle of the loop relying on the break statement after all the processing, which overall is a bit icky.